test/pipelinerun_test.go

```diff
@@ -35,175 +36,288 @@ import (
+				t.Fatalf("Failed to create SA `%s`: %s", getName(hwSA, index), err)
+			}
 
-	matchKinds := map[string]string{"PipelineRun": "", "TaskRun": ""}
-	events := collectMatchingEvents(t, c.KubeClient, namespace, matchKinds, "Succeeded")
-	if len(events) != expectedNumberOfEvents {
-		t.Errorf("Expected %d number of successful events from pipelinerun and taskrun but got %d", expectedNumberOfEvents, len(events))
-	}
-}
+			pp := getHelloWorldPipelineParams(index, namespace)
+			// Set SA pipeline params spec
+			pp.Spec.ServiceAccount = getName(hwSA, index)
+			if _, err := c.PipelineParamsClient.Create(pp); err != nil {
+				t.Fatalf("Failed to create PipelineParams `%s`: %s", pp.Name, err)
+			}
 
-// Test pipeline trun which refers to service account.
-func TestPipelineRun_WithServiceAccount(t *testing.T) {
-	logger := logging.GetContextLogger(t.Name())
-	c, namespace := setup(t, logger)
+			if _, err := c.TaskClient.Create(&v1alpha1.Task{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: namespace,
+					Name:      getName(hwTaskName, index),
+				},
+				Spec: v1alpha1.TaskSpec{
+					// Reference build: https://github.com/knative/build/tree/master/test/docker-basic
+					BuildSpec: &buildv1alpha1.BuildSpec{
+						Steps: []corev1.Container{
+							corev1.Container{
+								Name:  "config-docker",
+								Image: "gcr.io/cloud-builders/docker",
+								// Private docker image for Build CRD testing
+								Command: []string{"docker"},
+								Args:    []string{"pull", "gcr.io/build-crd-testing/secret-sauce"},
+								VolumeMounts: []corev1.VolumeMount{{
+									Name:      "docker-socket",
+									MountPath: "/var/run/docker.sock",
+								}},
+							},
+						},
+						Volumes: []corev1.Volume{{
+							Name: "docker-socket",
+							VolumeSource: corev1.VolumeSource{
+								HostPath: &corev1.HostPathVolumeSource{
+									Path: "/var/run/docker.sock",
+									Type: newHostPathType(string(corev1.HostPathSocket)),
+								},
+							},
+						}},
+					},
+				},
+			}); err != nil {
+				t.Fatalf("Failed to create Task `%s`: %s", getName(hwTaskName, index), err)
+			}
 
-	knativetest.CleanupOnInterrupt(func() { tearDown(t, logger, c, namespace) }, logger)
-	defer tearDown(t, logger, c, namespace)
+			if _, err := c.PipelineClient.Create(getHelloWorldPipelineWithSingularTask(index, namespace)); err != nil {
+				t.Fatalf("Failed to create Pipeline `%s`: %s", getName(hwPipelineName, index), err)
+			}
+		},
+		expectedTaskRuns:       []string{hwPipelineTaskName1},
+		expectedNumberOfEvents: 2,
+		// 1 from PipelineRun and 1 from Tasks defined in pipelinerun
+	}}
 
-	logger.Infof("Creating pipeline resources in namespace %s", namespace)
+	for i, td := range tds {
+		t.Run(td.name, func(t *testing.T) {
+			logger.Infof("Setting up test resources for %q test in namespace %s", td.name, namespace)
```

Comment:
if we want to run these tests in parallel, we should make sure to create a separate logger object for each test, b/c the output is going to be intermingled, and the named loggers will let us distinguish the lines from each other:

https://github.com/knative/build-pipeline/blob/a4a92776e8dbb371aa7d06742ce50f0e0769603e/test/init_test.go#L101-L109

Looks like I was doing this wrong from the start! what we really want is for the logger to be named after the test.

p.s. I think we're going to have some trouble with the metrics emitted by the tests, if we ever want to start using them, b/c [i'm pretty sure the metrics exporter uses a global logger](https://github.com/knative/pkg/blob/master/test/logging/logging.go#L101) and every time we initialize it, we clobber the previous one, i.e. all metrics will look like they are coming from one test (we should make a task to follow up on this so we don't forget maybe)

I think it would be worth simulating a failure and taking a look at what the log output looks like - sadly the integration tests are swallowing the logs when they succeed, so it's hard to see what they look like :(

We should consider adding something to the test README.md explaining that the log output will be intermingled now.

Reply:
IIUC, right now, each individual test case in `TestPipelineRun` is run sequentially.

You can always create a logger for testing in parallel with the test case name.
https://github.com/golang/go/issues/19280#issuecomment-298390151

Reply:
> so it's hard to see what they look like :(

You can add `-v` option to `go test` to see logs. No need to make the test fail to see logs.

`t.Run(name)` will update the sub test name to be `ParentTestName/subTestName`. In this example it will be `TestPipelineRun/multiple_tasks` or `TestPipelineRun/service_account_setup`. Logs will not be intermingled. 

> each individual test case in TestPipelineRun is run sequentially.

Thats true. This refactor uses common test setup but runs the test sequentially.

Reply:
I did consider using `t.Parallel()` but the problem I had with this approach was if one of the test completes earlier then it continues to cleanup namespace so other test fails with random errors.

Reply:
> You can add -v option to go test to see logs. No need to make the test fail to see logs.

That's true when you're running the tests locally but I'm talking about running on presubmit.

If you look at [our presubmit tests](https://github.com/knative/build-pipeline/blob/master/test/e2e-tests.sh#L74), we don't have a lot of control over how the tests are run.

[`go_test_e2e`](https://github.com/knative/test-infra/blob/master/scripts/e2e-tests.sh#L80) _does_ use `-v`, BUT unfortunately [`report_go_test`](https://github.com/knative/test-infra/blob/master/scripts/library.sh#L235) swallows the output unless the test fails.

Anyway long story short, I was just saying that when we look at the prow logs, since the test is succeeding, we don't get to see what the output looks like.

I think the fix here is simpler than I thought tho, I didn't notice the `logger := logging.GetContextLogger(t.Name())` at the beginning of the test, all we have to do is move that into the table driven part so we can use a different logger for each subtest.

> I did consider using t.Parallel() but the problem I had with this approach was if one of the test completes earlier then it continues to cleanup namespace so other test fails with random errors.

Ohhh okay, so these tests don't run in parallel after all.

Well we could update the test to create a namespace for each subtest?

Reply:
I created https://github.com/knative/build-pipeline/pull/211